src/app/clusters/mode-base-server/mode-base-server.cpp

```diff
@@ -0,0 +1,445 @@
+            if (emberAfContainsServer(mEndpointId, OnOff::Id) &&
+                emberAfContainsAttribute(mEndpointId, OnOff::Id, OnOff::Attributes::StartUpOnOff::Id) &&
+                emberAfContainsAttribute(mEndpointId, mClusterId, ModeBase::Attributes::OnMode::Id) &&
+                HasFeature(ModeBase::Feature::kOnOff))
+            {
+                DataModel::Nullable<uint8_t> onMode = GetOnMode();
+                bool onOffValueForStartUp           = false;
+                if (!emberAfIsKnownVolatileAttribute(mEndpointId, OnOff::Id, OnOff::Attributes::StartUpOnOff::Id) &&
+                    OnOffServer::Instance().getOnOffValueForStartUp(mEndpointId, onOffValueForStartUp) == EMBER_ZCL_STATUS_SUCCESS)
+                {
+                    if (onOffValueForStartUp && !onMode.IsNull())
+                    {
+                        ChipLogDetail(Zcl, "ModeBase: CurrentMode is overwritten by OnMode");
+                        // it is overwritten by the on/off cluster.
+                        return CHIP_NO_ERROR;
+                    }
+                }
+            }
+#endif // EMBER_AF_PLUGIN_ON_OFF
+
+            BootReasonType bootReason = BootReasonType::kUnspecified;
+            CHIP_ERROR error          = DeviceLayer::GetDiagnosticDataProvider().GetBootReason(bootReason);
+
+            if (error != CHIP_NO_ERROR)
+            {
+                ChipLogError(Zcl, "Unable to retrieve boot reason: %" CHIP_ERROR_FORMAT, error.Format());
+                // We really only care whether the boot reason is OTA.  Assume it's not.
+                bootReason = BootReasonType::kUnspecified;
+            }
+            if (bootReason == BootReasonType::kSoftwareUpdateCompleted)
+            {
+                ChipLogDetail(Zcl, "ModeBase: CurrentMode is ignored for OTA reboot");
+                return CHIP_NO_ERROR;
+            }
+
+            // Initialise currentMode to 0
+            uint8_t currentMode = GetCurrentMode();
+
+            if (startUpMode.Value() != currentMode)
+            {
+                if (!IsSupportedMode(startUpMode.Value()))
+                {
+                    ChipLogError(Zcl, "ModeBase: Start-up_mode is not set to a valid mode.");
+                    return CHIP_ERROR_INVALID_ARGUMENT; // todo is thins the correct error?
+                }
+
+                UpdateCurrentMode(startUpMode.Value());
+                ChipLogProgress(Zcl, "ModeBase: Successfully initialized CurrentMode to %u", startUpMode.Value());
+            }
+        }
+    }
+    else
+    {
+        ChipLogProgress(Zcl, "ModeBase: Skipped initializing CurrentMode by StartUpMode because one of them is volatile");
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+template <typename RequestT, typename FuncT>
+void Instance::HandleCommand(HandlerContext & handlerContext, FuncT func)
+{
+    if (!handlerContext.mCommandHandled && (handlerContext.mRequestPath.mCommandId == RequestT::GetCommandId()))
```

Comment:
When is the `if` condition ever false? When is `handlerContext.mCommandHandled` ever false?

Also, suggest rather to do an early exit:
```
if (handlerContext.mCommandHandled || (handlerContext.mRequestPath.mCommandId != RequestT::GetCommandId()) 
{
    return;
}
```